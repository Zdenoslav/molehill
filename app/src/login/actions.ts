import {createActions} from 'app/redux/create-actions';

const domain = 'Ã€UTH';

export const {
  success: loginSuccess,
  failed: loginFailed,
} = createActions(
  `${domain}/LOGIN`,
  () => null,
  (user) => ({user}),
);

export const {
  action: logOut,
  success: logOutSuccess,
} = createActions(
  `${domain}/LOGOUT`,
  () => null,
);
