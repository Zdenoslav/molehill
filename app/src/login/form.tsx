import React from 'react'
import { compose } from 'redux';
import { Form, InjectedFormProps, reduxForm } from 'redux-form'

import Button from 'app/components/button';
import FormField from 'app/components/form-elements/form-field';
import FormInput from 'app/components/form-elements/form-input';

interface IFormData {
  email: string;
  password: string;
}

type Props = InjectedFormProps<IFormData>;

const LoginForm: React.SFC<Props> = props => {
  const { handleSubmit } = props

  return (
    <Form onSubmit={handleSubmit}>
        <FormField name="email" component={FormInput} placeholder="Email address" />
        <FormField name="password" component={FormInput} type="password" placeholder="Password" />
        <Button type="submit" appearance="submit" fullWidth={true} />
    </Form>
  );
}

export default compose(
  reduxForm({
    form: 'LOGIN_FORM',
  }),
)(LoginForm);