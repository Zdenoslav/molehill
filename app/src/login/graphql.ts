import gql from 'graphql-tag';

export const loginMutation = gql`
  mutation login($loginInput: LoginInput!) {
    login(loginInput: $loginInput) {
      id
      username
      email
    }
  }
`;
