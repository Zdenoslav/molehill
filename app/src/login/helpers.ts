import { getFromLocalStorage, saveToLocalStorage } from 'app/helpers/local-storage';

export const getUserIdFromLocalStorage = () => getFromLocalStorage('userId');
export const saveUserIdToLocalStorage = (userId: string) => saveToLocalStorage('userId', userId);
