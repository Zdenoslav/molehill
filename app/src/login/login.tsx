import gql from 'graphql-tag';
import React from 'react';
import {graphql, MutateProps} from 'react-apollo';
import { Link } from 'react-router-dom';
import {compose} from 'redux';
import styled from 'styled-components';

import { s4 } from 'app/components/styled-components/spacing';

import ScreenCenter from 'app/components/screen-center';
import { Body, Headline } from 'app/components/styled-components/text';

import Form from './form';

const FormContainer = styled.div`
  margin-top: 16px;
`;

type Props = MutateProps;

const Login: React.SFC<Props> = (props) => {

  const onSubmit = (values: FormData) => {
    props.mutate({
     variables: {
      loginInput: values,
     },
    });
  };

  return (
    <ScreenCenter>
      <Headline textAlign="center">Molehill</Headline>
      <FormContainer>
        <Form onSubmit={onSubmit}/>
      </FormContainer>
      <Body marginTop={s4} textAlign="center">Don't have an account? <Link to="/signup">Sign up</Link></Body>
    </ScreenCenter>
  );
};

export default compose(
  graphql(gql`
    mutation login($loginInput: LoginInput!) {
      login(loginInput: $loginInput) {
        id
        username
        email
      }
    }
  `),
)(Login);
