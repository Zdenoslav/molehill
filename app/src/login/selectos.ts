import {createSelector} from 'reselect';

import { IRootState } from 'app/redux/root-reducer';

export const getUser = (state: IRootState) => state.login.user;

export const getUserId = createSelector(
  getUser,
  (user) => user ? user.id : '',
);

export const getUsername = createSelector(
  getUser,
  (user) => user ? user.username : '',
);
