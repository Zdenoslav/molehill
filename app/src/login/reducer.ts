import {Record} from 'immutable';

import { IReduxAction } from 'app/redux/create-actions';

import * as Actions from './actions';
import { Login_login } from 'app/generated/graphql';

export type User = Pick<Login_login, Exclude<keyof Login_login, '__typename'>>;
export class LoginState extends Record<{
  user: User | null,
}>({
  user: null,
}) {}

const loginReducer = (state = new LoginState(), action: IReduxAction) => {
  switch (action.type) {
    case Actions.loginSuccess.type:
      return state.merge({
        user: action.payload.user,
      });
    case Actions.logOutSuccess.type:
      return new LoginState();
    default:
      return state;
  }
};

export default loginReducer;
