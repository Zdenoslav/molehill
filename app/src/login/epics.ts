import { combineEpics, ofType, Epic } from 'redux-observable';
import { tap, ignoreElements } from 'rxjs/operators';

import {loginSuccess} from 'app/login/actions';
import { saveUserIdToLocalStorage } from './helpers';
import { IReduxAction } from 'app/redux/create-actions';

const storeUserIdInLocalStorage: Epic<IReduxAction<any>, any> = (action$) => action$.pipe(
  ofType(loginSuccess.type),
  tap((loginSuccessAction) => {
    saveUserIdToLocalStorage(loginSuccessAction.payload.user.id);
  }),
  ignoreElements(),
);

export default combineEpics(
  storeUserIdInLocalStorage,
);
