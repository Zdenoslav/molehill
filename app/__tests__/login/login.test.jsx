import React from 'react';
import { shallow } from 'enzyme';

import {Login} from 'app/login/login';

describe('Login', () => {

  it('should render Login', () => {
    const wrapper = shallow(<Login sMutation={{loading: false}} />);
    expect(wrapper).toMatchSnapshot();
  });

});
