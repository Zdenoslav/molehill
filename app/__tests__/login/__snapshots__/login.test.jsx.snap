// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render Login 1`] = `
<ScreenCenter>
  <Styled(Styled(styled.div))
    textAlign="center"
  >
    Molehill
  </Styled(Styled(styled.div))>
  <styled.div
    marginTop="16px"
  >
    <ReduxForm
      loading={false}
      onSubmit={[Function]}
    />
  </styled.div>
  <Styled(Styled(styled.div))
    marginTop="16px"
    textAlign="center"
  >
    Don't have an account? 
    <Link
      replace={false}
      to="/signup"
    >
      Sign up
    </Link>
  </Styled(Styled(styled.div))>
</ScreenCenter>
`;
